src/headers/lnurl_auth_jwt.rs

```diff
@@ -0,0 +1,299 @@
+			.get(&signed_lnurl)
+			.send()
+			.await
+			.map_err(HeaderProviderError::from)?
+			.json()
+			.await
+			.map_err(HeaderProviderError::from)?;
+
+		match lnurl_auth_response {
+			LnurlAuthResponse { token: Some(token), .. } => Ok(token),
+			LnurlAuthResponse { reason: Some(reason), .. } => {
+				Err(HeaderProviderError::ApplicationError(format!("LNURL Auth failed, reason is: {}", reason)))
+			}
+			_ => Err(HeaderProviderError::InvalidData(
+				"LNURL Auth response did not contain a token nor an error".to_string(),
+			)),
+		}
+	}
+
+	async fn get_jwt_token(&self, force_refresh: bool) -> Result<String, HeaderProviderError> {
+		if !self.is_expired() && !force_refresh {
+			let jwt_token = self.jwt_token.lock().unwrap();
+			if let Some(jwt_token) = jwt_token.as_deref() {
+				return Ok(jwt_token.to_string());
+			}
+		}
+		let jwt_token = self.fetch_jwt_token().await?;
+		let expiry = parse_expiry(&jwt_token)?;
+		*self.jwt_token.lock().unwrap() = Some(jwt_token.clone());
+		*self.expiry.lock().unwrap() = expiry;
+		Ok(jwt_token)
+	}
+
+	fn is_expired(&self) -> bool {
+		self.expiry
+			.lock()
+			.unwrap()
+			.map(|expiry| {
+				SystemTime::now().duration_since(SystemTime::UNIX_EPOCH).unwrap().as_secs() + EXPIRY_BUFFER_SECS
+					> expiry
+			})
+			.unwrap_or(false)
+	}
+}
+
+#[async_trait]
+impl HeaderProvider for LnurlAuthJwt {
+	async fn get_headers(&self) -> Result<HeaderMap, HeaderProviderError> {
+		let jwt_token = self.get_jwt_token(false).await?;
+		let mut headers = self.headers.clone();
+		let value = format!("Bearer {}", jwt_token).parse().map_err(HeaderProviderError::from)?;
+		headers.insert(AUTHORIZATION, value);
+		Ok(headers)
+	}
+}
+
+fn hashing_key(engine: &Secp256k1<All>, parent_key: &ExtendedPrivKey) -> Result<PrivateKey, HeaderProviderError> {
+	let hashing_child_number =
+		ChildNumber::from_normal_idx(HASHING_DERIVATION_INDEX).map_err(HeaderProviderError::from)?;
+	parent_key
+		.derive_priv(engine, &vec![hashing_child_number])
+		.map(|xpriv| xpriv.to_priv())
+		.map_err(HeaderProviderError::from)
+}
+
+fn linking_key_path(hashing_key: &PrivateKey, domain_name: &str) -> Result<DerivationPath, HeaderProviderError> {
+	let mut engine = HmacEngine::<sha256::Hash>::new(&hashing_key.inner[..]);
+	engine.input(domain_name.as_bytes());
+	let result = Hmac::<sha256::Hash>::from_engine(engine).to_byte_array();
+	let children: Vec<ChildNumber> = (0..4)
+		.map(|i| u32::from_be_bytes(result[(i * 4)..((i + 1) * 4)].try_into().unwrap()))
+		.map(ChildNumber::from)
+		.collect::<Vec<_>>();
+	Ok(DerivationPath::from(children))
+}
+
+fn sign_lnurl(
+	engine: &Secp256k1<All>, parent_key: &ExtendedPrivKey, lnurl_str: &str,
+) -> Result<String, HeaderProviderError> {
+	// Parse k1 parameter to sign.
+	let invalid_lnurl = || HeaderProviderError::InvalidData(format!("invalid lnurl: {}", lnurl_str));
+	let mut lnurl = Url::parse(lnurl_str).map_err(|_| invalid_lnurl())?;
+	let domain = lnurl.domain().ok_or(invalid_lnurl())?;
```

Comment:
Yes, exactly, the key derivation and signature will depend on it, likely to avoid certain attacks. I wonder if we should enforce HTTPS as well.

Reply:
@G8XSU @tnull Were there remaining questions here from the LDK sync?

From my understanding, the domain name is used for isolation so that a signature for an LNURL Auth-based todo app for example cannot be used to access the LNURL Auth-based VSS server. HTTPS also protects against man-in-the-middle attacks.